### PR TITLE
Further fixes and updates for tf-head

### PR DESCRIPTION
**Pull request by maxhgerlach**
## Description

Let's collect more fixes for TensorFlow nightly builds here.

### CODE REVIEW

**Comment by maxhgerlach** [comment]
There's a couple of opimizer-related errors in `test_spark_keras.py` with tf-nightly:

- `test_fit_model`
- `test_fit_model_multiclass`
- `test_keras_direct_parquet_train`
- `test_keras_model_checkpoint_callback`
- `test_model_serialization`
- `test_restore_from_checkpoint`

(see collected tracebacks in https://github.com/horovod/horovod/runs/12877848393)

I don't have a good understanding about how the optimizers are wrapped there. @chongxiaoc, do you maybe have an idea how to fix these?

**Comment by maxhgerlach** [comment]
`DistributedOptimizer.variables` is expected to be a property from Keras 2.12, which is behind some of the Spark Keras failures. Let's see how my attempted fix fares.

**Comment by maxhgerlach** [comment]
All TensorFlow/Keras tests pass now.

There's a failure in Spark Lightning MNIST on the CPU/GPU head pipelines. It should be investigated separately.

File: horovod/_keras/__init__.py [hunk 1]
```diff
@@ -84,6 +84,9 @@ def variables(self):
                 return super(self.__class__, self).variables()
             return self.get_weights()
 
+        if version.parse(tf.__version__) >= version.parse('2.12.0'):
```
Comment (nvcastet) [inline horovod/_keras/__init__.py:87]: Thanks @maxhgerlach I could figure out this one.
So TF changed the attribute from method to property? :)

**Review by nvcastet — APPROVED** [review]

File: horovod/_keras/__init__.py (hunk 1 above)
Comment (maxhgerlach) [reply horovod/_keras/__init__.py:87]: They did, yes: https://github.com/keras-team/keras/commit/50e0b18ceeb370157f36a387115c5ae1a16fca47

There's a little compatibility shim so code using an optimizer still works with both `.variables` and with `.variables()`.

Cheers!

**Pull request merged**